proposals/0000-exception-backtraces.rst

```diff
@@ -0,0 +1,553 @@
+annotations: ::
+
+    -- | A backtrace from the cost-centre profiler.
+    data CostCentreBacktrace = CostCentreBacktrace { ... }
+    instance ExceptionAnnotation CostCentreBacktrace
+
+    -- | A backtrace using the native (e.g. DWARF-based) stack unwinder
+    data ExecutionStackBacktrace = ExecutionStackBacktrace { ... }
+    instance ExceptionAnnotation ExecutionStackBacktrace
+
+    -- | A backtrace from GHC's Haskell stack unwinder and info-table
+    -- provenance map.
+    data InfoProvBacktrace = InfoProvBacktrace { ... }
+    instance ExceptionAnnotation InfoProvBacktrace
+
+    -- | A backtrace from HasCallStack evidence.
+    data HasCallStackBacktrace = HasCallStackBacktrace { ... }
+    instance ExceptionAnnotation HasCallStackBacktrace
+
+Handling of rethrowing
+^^^^^^^^^^^^^^^^^^^^^^
+
+One pattern frequently seen in Haskell programs is *rethrowing*. Typically this
+takes the form of catching one type of exception and throwing in its place
+another exception more specific to the application domain. For instance, ::
+
+    data MyAppError = MissingConfigurationError | ...
+
+    readFile "my-app.conf" `catch` $ \ (ioe :: IOError) ->
+        if isDoesNotExistError ioe
+          then throwIO MissingConfigurationError
+          else throwIO ioe
+
+This pattern can be problematic in the presence of exception context: the
+exception thrown by the handler lacks any of the context attached to the
+original ``IOError``, including any backtraces.
+
+While in some select cases dropping context may be desireable (e.g. to avoid
+exposing implementation details unnecessarily to the user), in general this
+proposal seeks to make exception provenance information ubiquitous and
+reliable. Consequently, we propose to that ``catch`` and ``handle`` be modified
+to preserve exception context when an exception is thrown from a handler.
+
+
+Selecting backtrace mechanisms
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+With the machinery described above, we can now address a common debugging
+scenario: locating the origin of an exception thrown by a third-party library.
+By far the most common means of throwing exceptions are ``throw``, ``throwIO``,
+``error``, and ``undefined``, none of which have any notion of backtrace collection.
+This raises the question of how the user should select which of the above
+mechanism(s) these functions should use to collect their backtrace.
+
+For this we propose a pragmatic, stateful approach to allow the user to enable
+individual mechanism(s) should be used for backtrace collection in ``throw``,
+``throwIO`` and similar functions: ::
+
+    module GHC.Exception.Backtrace
+        ( enableMechanism, BacktraceMechanism(..) ) where
+
+    -- | Which kind of backtrace to collect when an exception is thrown.
+    data BacktraceMechanism
+      = -- | collect a cost center stacktrace (only available when built with profiling)
+        CostCenterBacktraceMech
```

Comment:
I would spell out `BacktraceMechanism` rather than slightly abbreviating it in the constructor names, personally.